Adding the improved dot file generator feature.

alacasta opened the pull request:
Changelog: (Feature): This pull request contains an improved version of the `dot` output generation for the dependency graph. It cosiders the following facts
- Each node is connected as much as 1 times with any other node (the graph is not `strict`). 
- Each dependency is colorized in blue (following Conan colors)
- The root of the graph is colorized in yellow. 
- The edges are collocated improving the readability (not the size of the plot)

Apart from this, some additional stylish settings have been defined
- The nodes are a little bit separated (better reading)
- The font is less 'coding-style' 

Considering the following `conanfile.py`:

```python
from conans import ConanFile, CMake

class MyPkg(ConanFile):
    name = "mypkg"
    version = "1.2.3-rc1"
        
    def requirements(self):
        self.requires("openexr/2.4.0")
        self.requires("libjpeg/9d")
        self.requires("leptonica/1.78.0")
        self.requires("openssl/1.1.1f")
        self.requires("zlib/1.2.11")
        self.requires("libtiff/4.1.0")
        self.requires("cimg/2.8.3")
        self.requires("folly/2019.10.21.00")
        self.requires("zstd/1.4.3")
        self.requires("opencv/4.1.1@conan/stable")
        
    def build_requirements(self):
        self.build_requires("nasm/2.14")
        self.build_requires("gtest/1.8.1", force_host_context=True)  # 'host' context (our library will link with it
```

The original output when running

```
conan info . --graph=deps.dot
dot deps.dot -Tpng -o deps.png
```
was
![image](https://user-images.githubusercontent.com/1688725/78689008-28cc7f80-78f6-11ea-9d52-8807b610c205.png)

and the improved one is 

![image](https://user-images.githubusercontent.com/1688725/78690200-72699a00-78f7-11ea-96b3-d362e5feec0d.png)

## Additional changes

Apart from the changes indicated above, additional changes in collaboration with my colleague @saimusdev, which was also interested on this feature has been added.
- A clear differenciation based on colors for the following items
  - The package for which the diagraph is going to be plotted (in green)
  - The Build_Deps (in yellow)
- Some reordering of the nodes to have always 100% reproducible graphs. It was observed that from one ran to another, the graph was not estrictly the same. The origin for this was that the inner graph is not always created following the same ordering, and therfore, the dot file also changed. To avoid this
  - The nodes are ordered (alphabetically)
  - The dependencies are ordered (alphabetically)
- Improved documentation in the modified class. 

With this latest changes, the Diagraph shown below looks now like

![image](https://user-images.githubusercontent.com/1688725/78979906-15e7c400-7b1d-11ea-9937-c22e7b3392de.png)

Closes: #6814
Docs: This contribution has no reflection in the documentation because i found no (graphical) reference to it. Because this change is only an aesthetic change in the generated dot file, i consider not needed to create nothing special on this.

- [x] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

alacasta committed:
Adding the improved dot file generator feature.

alacasta committed:
Setting the colors to be used in the Dot file to those used by Conan (blue)

alacasta committed:
Reducing h-space for the nodes.

alacasta commented:
Hello @jgsogo!

I see many further ways for improving this, especially regarding the build/host contexts. However and since i am not really familiarized with all the code, i didn't see how to manage this in the graph (but i see a `#ToDo` remark on 
https://github.com/conan-io/conan/blob/cf4cab6f9fc95ada3b69796f6de615b4d79ed1ff/conans/client/graph/grapher.py#L76 
That indicated me that you already though on this :)

I think that this is a good first step that for sure can be improved in the future providing the context in this view. Unluckly i saw no way for introducing labels in the edges to indicate if it was 'require' or ' build require' in a cool way. All the labels was displaced and the result was not that good. 

I hope you find the contribution good enough to be considered to be included in some near new *Conan* release!

alacasta commented:
Jenkins is crying until I fix all the tests... The resulting file is radically different and the tests should be redesigned (or at least review). 

I will take care of them as soon as possible.

jgsogo commented:
Thanks a lot for the PR!

Let's talk first about the changes and then we can take care of the tests.

alacasta committed:
Changing to a more-standard and SVG supported font: Helvetica.

alacasta commented:
Dear Reviewer, 
The pull request is still in draft mode until the tests are fixed. Nevertheless, i'd appreciate a gross review on the changes/feature before working on them.
Thanks!

alacasta committed:
Switching back to the non-custom nodes labelling. It is possible to handle them as they were but only especify the format for each node (the current contribution)

alacasta commented:
The final contribution considers no modification in the nodes labelling (therefore, the tests should pass  nonetheless) 🎉 

I missed that `DOT` was not requiring to re-labelling them but you can use any string (i.e. `"pkg/0.0.1@conan/testing"`) as an identifier. By default, if any additional info is provided, it is used also as a value for the node. In turn and with the current approach, you can use it just for labelling the node while the value of the node, is especified beforehand.

alacasta commented:
Luckly tests now pass! ✔️ 

However, Jenkins is throwing a false negative bc os some internal issue

```
PermissionError: [Errno 13] Permission denied: 'D:\\J\\t\\d515\\py36\\disabled_revisions\\venv\\Scripts\\python.exe'
```

I'm gonna close and reopen the PR to re-trigger the CI

alacasta committed:
Reflexting the 'stricness' of the graph into the tests.

alacasta committed:
Fixing tests adding the new dot fields to the verification. Improving the file format.

alacasta commented:
Contribution ready to be reviewed. Additional (light) verifications on the dot output files.

saimusdev committed:
update client/graph/grapher.py (dot grapher)
selectively colors nodes depending if they're build_requires or normal requires

saimusdev committed:
update client/graph/grapher.py (dot grapher)
root node is colored in a different color (green)

saimusdev committed:
update client/graph/grapher.py (dot grapher)
nodes are gathered and then exported to the dot graph in order (by node name)

saimusdev committed:
update client/graph/grapher.py (dot grapher)
dependency matrix is also orderede & deterministic now

alacasta committed:
Comformming Pep8 format

alacasta committed:
Adapting Unit Tests

alacasta committed:
Re-defining the labels to the display_name to be compatible with the previous definition of the nodes. Reverting a minor change in a test to be consisten with this change. Removing the node colors from the global node definition.

alacasta committed:
Improving documentation of newer (and existing) functions.

alacasta committed:
Merge pull request #1 from alacasta/feature/6814-additional-features

Feature 6814: Additional Features

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/conan-io/conan?pullRequest=6815) <br/>All committers have signed the CLA.

alacasta commented:
@saimusdev  Please, sign-in the CLA clause.

alacasta committed:
Improving the dependencies addressing within the adjacency list.

alacasta committed:
Fixing the coloring of the root node.

alacasta commented:
@jgsogo / Reviewer:
I've realized that the html output is strongly different, even using the same sort algo (UD).
![image](https://user-images.githubusercontent.com/1688725/78980720-c86c5680-7b1e-11ea-9e4a-3d0780b6b0c8.png)

![image](https://user-images.githubusercontent.com/1688725/78980900-213bef00-7b1f-11ea-98cb-b866ecec7c9b.png)

I think that this is not about the contribution, but i am unsure if the resulting `html` output should look like the one generated using `dot` (i saw that for simple ones, they match, but not in this case).

jgsogo commented:
Hi, @alacasta 

Thanks a lot for the hard work. I'm figuring out how to add these changes to Conan and satisfy other users that might find a different design/coloring/shapes better. They can also prefer the default Graphviz render output 🤷 .

I did a couple of PRs during these days to add more flexibility to the `search --table` output and to the `info --graph` one, so the user can override a default Conan template with the one they like. With those changes templates are easily interchangeable and we can maintain a repository with user-contributed templates (Conan default, yours,... and I hope that someone else will want to contribute).

I'm asking you for a bit of patience, I need to introduce this new development to the team so we all are aligned 

Thanks!

alacasta commented:
> Hi, @alacasta
> 
> Thanks a lot for the hard work. I'm figuring out how to add these changes to Conan and satisfy other users that might find a different design/coloring/shapes better. They can also prefer the default Graphviz render output 🤷 .
> 
> I did a couple of PRs during these days to add more flexibility to the `search --table` output and to the `info --graph` one, so the user can override a default Conan template with the one they like. With those changes templates are easily interchangeable and we can maintain a repository with user-contributed templates (Conan default, yours,... and I hope that someone else will want to contribute).
> 
> I'm asking you for a bit of patience, I need to introduce this new development to the team so we all are aligned
> 
> Thanks!

Sounds good @jgsogo!

Thanks for the feedback and the notification 👍

saimusdev commented:
Hey, 

First of all thanks for your quick response! 
That sounds really cool. do you mean having HTML or whatever templates with keywords, to format, e.g. `<li>%NODE_NAME%</li>` ?

I've a question unrelated to your last message, but that has been for a while on the back of my mind... Would it be somewhat possible to include to the `graph` function, as an input, or in any other way, the `python_requires` dependencies?
It would be awesome to include those in the generated dotfile, and if possible and even awesomer, in a different color, using some variable, as we did with the `build_requires`.

What do you think?

thanks!

saimusdev commented:
Hi @jgsogo !

Any updates?
Could you answer our last post?

Thanks!

jgsogo commented:
Hi! PR https://github.com/conan-io/conan/pull/6833 has been merged. It, together with the refactor to the search table output, are two steps toward using templates for these outputs. 

As you could imagine we cannot make an output that would satisfy all our users, there will be always someone who would love to change a color or the way we display some data, so the best we can do IMHO is to provide a default output, and let the user override it using a template syntax with a clear interface.

Now we are using templates for these files, you can find them in this directory (https://github.com/conan-io/conan/tree/develop/conans/assets/templates), they are using jinja2 syntax, which is really powerful. Those templates (the graph ones) are feed with a [`grapher` object](https://github.com/conan-io/conan/blob/develop/conans/client/graph/grapher.py) that contains all the information from the Conan graph.

Next steps could be (TBD):
 * make it possible to use user-defined templates, probably they should be stored in a `templates` directory inside the Conan cache, and Conan will look for them (by name) before falling back to the default ones.
 * document objects and their attributes/functions available for the template files.

On top of this:
 * more data can be added to those objects (python_requires, build_requires) and it will be up to the template to show that data or not and to choose the color/shape for those nodes.
 * I would love to see the community sharing their templates, especially the graph one, which is something that can we really cool with a little bit of javascript.

I hope this covers most requests related to this output. It is not going as fast as all of us would love to, but there are a lot of things under development right now 🧑‍🏭

alacasta commented:
Thanks @jgsogo,

I find your new approach more convenient for the long term, so i have nothing to add there.

I think that we do still need something better than the default output, but for sure it is required first to open a mechanism to do that. I don’t know if we can contribute on it but if there is some open Issue documented enough, perhaps we can try to help you to make it possible sooner.

I hope we can see this feature soon in the always amazing Conan tool! 😊

jgsogo commented:
Merged! https://github.com/conan-io/conan/pull/7176

[Docs are ready here](https://github.com/conan-io/docs/pull/1739) (and approved), so we have public-experimental interface to implement other templates.

Given these changes I'm not sure if this can be a PR to Conan itself or it is more like a community contribution we can (and want) link to from the docs. For the `graph.dot` output I prefer to keep the raw Graphviz render, it is something nobody likes, but everyone agrees with it.

The way to go would be to provide this new template in a URL (another Github repo? a gist?) and we will add to the docs this one as an example to install a different template with `conan config install .... --target-folder ...` (and a nice image).

What do you think? Is it ok for you? I can help with the jinja2 template if you are not familiarized with its syntax.

Thanks!

alacasta commented:
> Merged! https://github.com/conan-io/conan/pull/7176
> 
> 
> 
> [Docs are ready here](https://github.com/conan-io/docs/pull/1739) (and approved), so we have public-experimental interface to implement other templates.
> 
> 
> 
> Given these changes I'm not sure if this can be a PR to Conan itself or it is more like a community contribution we can (and want) link to from the docs. For the `graph.dot` output I prefer to keep the raw Graphviz render, it is something nobody likes, but everyone agrees with it.
> 
> 
> 
> The way to go would be to provide this new template in a URL (another Github repo? a gist?) and we will add to the docs this one as an example to install a different template with `conan config install .... --target-folder ...` (and a nice image).
> 
> 
> 
> What do you think? Is it ok for you? I can help with the jinja2 template if you are not familiarized with its syntax.
> 
> 
> 
> Thanks!

Thanks @jgsogo 

I’ve seen that the feature has been finally merged!! 😊

I am going to see how can I adapt this to the new template format. I do agree that this approach is much more flexible and may bring really nice output templates from the community. 

I hope I can work on this soon!

alacasta commented:
> Merged! #7176
> 
> [Docs are ready here](https://github.com/conan-io/docs/pull/1739) (and approved), so we have public-experimental interface to implement other templates.
> 
> Given these changes I'm not sure if this can be a PR to Conan itself or it is more like a community contribution we can (and want) link to from the docs. For the `graph.dot` output I prefer to keep the raw Graphviz render, it is something nobody likes, but everyone agrees with it.
> 
> The way to go would be to provide this new template in a URL (another Github repo? a gist?) and we will add to the docs this one as an example to install a different template with `conan config install .... --target-folder ...` (and a nice image).
> 
> What do you think? Is it ok for you? I can help with the jinja2 template if you are not familiarized with its syntax.
> 
> Thanks!

Hi @jgsogo !

I have finally converted this contribution into a `jinja2` template as you suggested 🎉 and I think that we can close this PR which makes no sense now. 

https://github.com/alacasta/beautified-conan-info-dot-template

The API available is enough for displaying basic information although further fields may result in highly enriched representations. I've opened an issue abouth this https://github.com/conan-io/conan/issues/7354 just in case you consider it as well.

I think that we can expect cool contribs from the community with even better templates than the one proposed by me... 

Thanks!

cc/ @saimusdev

alacasta commented:
@jgsogo Btw:

If there is any available mechanism provided to make this also visible to the hole community, just let me know how and I will try to proceed if you think that the template is good enough. 😉

jgsogo commented:
Thank you very much! 

I have one next movement in mind to make all these contributions visible to the whole community. I'll ping you when it is ready, I think I will be able to work on it soon... after my holidays 🌴
